algorithms/nautilus/grammartec/tree.cpp

```diff
@@ -0,0 +1,618 @@
+ */
+Tree TreeMutation::ToTree(Context& ctx) const {
+  std::vector<RuleIDOrCustom> vec;
+  vec.insert(vec.end(), _prefix.begin(), _prefix.end());
+  vec.insert(vec.end(), _repl.begin(), _repl.end());
+  vec.insert(vec.end(), _postfix.begin(), _postfix.end());
+  return Tree(std::move(vec), ctx);
+}
+
+/**
+ * @fn
+ * Get the rule of a node.
+ * @brief Get rule by node ID
+ * @param (n) Node ID
+ * @param (ctx) Context
+ * @throw std::out_of_range Node ID is invalid
+ * @return Rule corresponding to node ID
+ */
+const Rule& TreeMutation::GetRule(const NodeID& n, Context& ctx) const {
+  return ctx.GetRule(GetRuleID(n));
+}
+
+/**
+ * @fn
+ * Get the RuleIDOrCustom of a node.
+ * @brief Get rule ID or custom by node ID
+ * @param (n) Node ID
+ * @throw std::out_of_range Node ID is invalid
+ * @return RuleIDOrCustom corresponding to node ID
+ */
+const RuleIDOrCustom& TreeMutation::GetRuleOrCustom(const NodeID& n) const {
+  return GetAt(n);
+}
+
+/**
+ * @fn
+ * Get the custom rule data of a node.
+ * @brief Get custom rule data by node ID
+ * @param (n) Node ID
+ * @throw std::out_of_range Node ID is invalid
+ * @return Data of rule (throws exception if rule is not Custom)
+ */
+const std::string& TreeMutation::GetCustomRuleData(const NodeID& n) const {
+  return GetAt(n).Data();
+}
+
+/**
+ * @fn
+ * Get the nonterminal ID of a node.
+ * @brief Get nonterminal ID by NodeID
+ * @param (n) NodeID
+ * @param (ctx) Context
+ * @throw std::out_of_range Node ID is invalid
+ * @return Nonterminal symbol ID
+ */
+const NTermID& TreeLike::GetNontermID(const NodeID& n, Context& ctx) const {
+  return GetRule(n, ctx).Nonterm();
+}
+
+/**
+ * @fn
+ * Unparse (sub)tree derived from a node into a testcase.
+ * @brief Unparse a subtree into string
+ * @param (id) Node ID
+ * @param (ctx) Context
+ * @param (data) Reference to string to store the result
+ */
+void TreeLike::Unparse(const NodeID& id, Context& ctx, std::string& data) const {
+  Unparser(id, data, *this, ctx).Unparse();
+}
+
+/**
+ * @fn
+ * Unparse the whole tree into a testcase.
+ * @brief Unparse this tree into string
+ * @param (ctx) Context
+ * @param (data) Reference to string to store result
+ */
+void TreeLike::UnparseTo(Context& ctx, std::string& data) const {
+  Unparse(NodeID(0), ctx, data);
+}
+
+std::string TreeLike::UnparseNodeToVec(const NodeID& n, Context& ctx) const {
+  std::string data;
+  Unparse(n, ctx, data);
+  return data;
+}
+
+std::string TreeLike::UnparseToVec(Context& ctx) const {
+  return UnparseNodeToVec(NodeID(0), ctx);
+}
+
+
+/**
+ * @fn
+ * Construct an Unparser instance.
+ * @brief Construct Unparser
+ * @param (nid) Node ID
+ * @param (w) Data
+ * @param (tree) Tree
+ * @param (ctx) Context
+ */
+Unparser::Unparser(
+  const NodeID& nid, std::string& w, const TreeLike& tree, Context& ctx
+) : _tree(tree), _w(w), _ctx(ctx) {
+  _i = static_cast<size_t>(nid);
+  _stack.emplace_back(tree.GetRule(NodeID(_i), ctx).Nonterm());
+  _buffers.clear();
```

Comment:
`_buffers` is originally empty, isn't it?

Reply:
Fixed in [da9dbdb](https://github.com/fuzzuf/fuzzuf/pull/23/commits/da9dbdbabb27b0647b085c9a8d471a7328ee2e37)